File: ganmenScouter_android/src/main/java/com/ganmen/GanmenScouter.java
package com.ganmen;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Bitmap.CompressFormat;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.Matrix;
import android.hardware.Camera;
import android.hardware.Camera.Size;
import android.os.Bundle;
import android.provider.MediaStore;
import android.provider.MediaStore.Images;
import android.view.MotionEvent;
import android.view.Surface;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout.LayoutParams;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.InterstitialAd;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.initialization.InitializationStatus;
import com.google.android.gms.ads.initialization.OnInitializationCompleteListener;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.List;
/*
import com.ganmen.AnalyticsApplication;
import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.Tracker;
import com.google.android.gms.security.ProviderInstaller.ProviderInstallListener;
*/

public class GanmenScouter extends Activity {
	private final int WRAP_CONTENT = ViewGroup.LayoutParams.WRAP_CONTENT;
	static final int REQUEST_CODE = 1;

    // プレビューサイズ
    static private int width = -1;
    static private int height = -1;
    
	private Button mchange_btn;
	private Button incam_btn;

	private TextView tv_top;
	
    // カメラインスタンス
    private Camera mCam = null;
    private Camera inCam = null;

    // カメラプレビュークラス
    private CameraPreview mCamPreview = null;

    // 画面タッチの2度押し禁止用フラグ
    private boolean mIsTake = false;
    
    private boolean isMaleMode = false;
    private boolean isInCamMode = false;
    
    View rootView_ = null;
    
    Activity root_act = this; 
    
    boolean called_intent = false;
 
	InterstitialAd mInterstitialAd;
    
    RelativeLayout preview;

    FacePPService facePP = new FacePPService();

    public static GanmenScouter curentInstance = null;

    private boolean hasInCam(){
    	int numberOfCameras = Camera.getNumberOfCameras();
    	if(numberOfCameras >= 2){
    		return true;
    	}else{
        	return false;    		
    	}
    }

    @Override
    public void onStart() {
      super.onStart();
      GanmenScouter.curentInstance = this;
	}

    private void getPermissions(){
		if (ContextCompat.checkSelfPermission(this,
				Manifest.permission.CAMERA)
				!= PackageManager.PERMISSION_GRANTED) {


			if (ActivityCompat.shouldShowRequestPermissionRationale(this,
					Manifest.permission.CAMERA)) {
				(new AlertDialog.Builder(this)).setMessage("端末の「設定」でカメラの権限を許可してください。")
						.setPositiveButton("OK", new DialogInterface.OnClickListener() {
							public void onClick(DialogInterface dialog, int id) {
							}
						})
						.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
							public void onClick(DialogInterface dialog, int id) {
							}
						}).create().show();
			} else {
				ActivityCompat.requestPermissions(this,
						new String[]{Manifest.permission.CAMERA}, REQUEST_CODE);
			}
		} else {
		}

		if (ContextCompat.checkSelfPermission(this,
				Manifest.permission.WRITE_EXTERNAL_STORAGE)
				!= PackageManager.PERMISSION_GRANTED) {

			if (ActivityCompat.shouldShowRequestPermissionRationale(this,
					Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
				(new AlertDialog.Builder(this)).setMessage("端末の「設定」でストレージへの書き込みの権限を許可してください。")
						.setPositiveButton("OK", new DialogInterface.OnClickListener() {
							public void onClick(DialogInterface dialog, int id) {
							}
						})
						.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
							public void onClick(DialogInterface dialog, int id) {
							}
						}).create().show();
			} else {
				ActivityCompat.requestPermissions(this,
						new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_CODE);
			}
		} else {
		}
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

        setContentView(R.layout.main);
		getPermissions();

		MobileAds.initialize(this, new OnInitializationCompleteListener() {
			@Override
			public void onInitializationComplete(InitializationStatus initializationStatus) {
				requestNewInterstitial();
			}
		});

		mInterstitialAd = new InterstitialAd(this);
		mInterstitialAd.setAdUnitId("ca-app-pub-3869533485696941/9899777318");

		mInterstitialAd.setAdListener(new AdListener() {
			@Override
			public void onAdClosed() {
				requestNewInterstitial();
				setup_cam_and_preview(false);
			}
		});

        setup_cam_and_preview(true);

		preview = (RelativeLayout) findViewById(R.id.cameraPreview);

		tv_top = new TextView(this);
		tv_top.setTextColor(Color.RED);
		tv_top.setText("　　　　　　　　　　　　　　　画面タッチで測定!");
		preview.addView(tv_top, LayoutParams.WRAP_CONTENT);

		incam_btn = (Button) findViewById(R.id.incam_btn);
		if (hasInCam()) {
			incam_btn.setText("インカムモードへ切替");
			incam_btn.setOnClickListener(new OnClickListener() {
				public void onClick(View v) {
					if (isInCamMode == false) {
						isInCamMode = true;
//						incam_btn.setText("背面カメラモードへ切替");
						incam_btn.setVisibility(View.INVISIBLE);
						//setup_in_cam_and_preview();
						setup_cam_and_preview(false);
					}
				}
			});
		} else {
			incam_btn.setText("インカム無し");
			incam_btn.setVisibility(View.INVISIBLE);
		}
    }
    
    private void requestNewInterstitial() {
        mInterstitialAd.loadAd(new AdRequest.Builder().build());
    }

	private void setup_cam_and_preview(boolean is_first){
		if(is_first==false){
			RelativeLayout preview = (RelativeLayout) findViewById(R.id.cameraPreview);
			preview.removeView(mCamPreview);
		}

		// カメラインスタンスの取得
		try {
			if(mCam!=null){
				mCam.stopPreview();
				mCam.release();
				mCam = null;
			}
			if(inCam!=null){
				inCam.stopPreview();
				inCam.release();
				inCam = null;
			}
			if(isInCamMode && inCam == null){
				inCam = Camera.open(1);
			}else if(mCam == null){
				mCam = Camera.open(0);
			}
		} catch (Exception e) {
			// エラー
			this.finish();
		}

		final Camera currentCam = isInCamMode ? inCam : mCam;

		Camera.Parameters params = currentCam.getParameters();

		int min_width = -1;
		//Picture解像度取得
		List<Size> supportedPictureSizes = params.getSupportedPictureSizes();
		if (supportedPictureSizes != null && supportedPictureSizes.size() > 1) {
			for(int i=0;i<supportedPictureSizes.size();i++){
				Size size = supportedPictureSizes.get(i);
				//解像度表示
				System.out.println("picture width:"+size.width+" height:"+size.height);
				if(min_width == -1){
					width = size.width;
					height = size.height;
				}else if(size.width < width){
					width = size.width;
					height = size.height;
				}
			}
		}

		params.setPictureSize(width, height);
		params.setPreviewSize(width, height);

		RelativeLayout preview = (RelativeLayout) findViewById(R.id.cameraPreview);

		mCamPreview = new CameraPreview(this, currentCam);
		if(isInCamMode){
			setCameraDisplayOrientation(this, 1, inCam);
		}else{
			setCameraDisplayOrientation(this, 0, mCam);
		}

		preview.addView(mCamPreview);

		// mCamPreview に タッチイベントを設定
		mCamPreview.setOnTouchListener(new View.OnTouchListener() {
			public boolean onTouch(View v, MotionEvent event) {
				if (event.getAction() == MotionEvent.ACTION_DOWN) {
					if (!mIsTake) {
						// 撮影中の2度押し禁止用フラグ
						mIsTake = true;
						if (currentCam != null) {
							// 撮影実行(AF開始)
							currentCam.autoFocus(autoFocusListener_);
						}
						// 画像取得
						// mCam.takePicture(null, null, mPicJpgListener);
					}
				}
				return true;
			}
		});

		tv_top = new TextView(this);
		tv_top.setTextColor(Color.RED);
		tv_top.setText("　　　　　　　　　　　　　　　画面タッチで測定!");
		preview.addView(tv_top, LayoutParams.WRAP_CONTENT);

		mchange_btn = new Button(this);
		mchange_btn.setText("男性撮影モードへ切替");
		mchange_btn.setOnClickListener(new OnClickListener() {
			public void onClick(View v) {
				if(isMaleMode == false){
					isMaleMode = true;
					mchange_btn.setText("女性撮影モードへ切替");
				}else{
					isMaleMode = false;
					mchange_btn.setText("男性撮影モードへ切替");
				}
			}
		});
		preview.addView(mchange_btn);
	}

	 public static void setCameraDisplayOrientation(Activity activity,
	         int cameraId, android.hardware.Camera camera) {
	     android.hardware.Camera.CameraInfo info =
	             new android.hardware.Camera.CameraInfo();
	     android.hardware.Camera.getCameraInfo(cameraId, info);
	     int rotation = activity.getWindowManager().getDefaultDisplay()
	             .getRotation();
	     int degrees = 0;
	     switch (rotation) {
	         case Surface.ROTATION_0: degrees = 0; break;
	         case Surface.ROTATION_90: degrees = 90; break;
	         case Surface.ROTATION_180: degrees = 180; break;
	         case Surface.ROTATION_270: degrees = 270; break;
	     }

	     int result;
	     if (info.facing == Camera.CameraInfo.CAMERA_FACING_FRONT) {
	         result = (info.orientation + degrees) % 360;
	         result = (360 - result) % 360;  // compensate the mirror
	     } else {  // back-facing
	         result = (info.orientation - degrees + 360) % 360;
	     }
	     camera.setDisplayOrientation(result);
	 }

    // AF完了時のコールバック
    private Camera.AutoFocusCallback autoFocusListener_ = new Camera.AutoFocusCallback() {
        @Override
        public void onAutoFocus(boolean success, Camera camera) {
            camera.autoFocus(null);
            if(isInCamMode){
            	inCam.takePicture(null, null, mPicJpgListener);            	
            }else{
            	mCam.takePicture(null, null, mPicJpgListener);	
            }
        }
    };

    /**
     * JPEG データ生成完了時のコールバック
     */
    private Camera.PictureCallback mPicJpgListener = new Camera.PictureCallback() {
    	
        public void onPictureTaken(byte[] data, Camera camera) {
            if (data == null) {
                return;
            }
            
            Bitmap origBitmap = BitmapFactory.decodeByteArray(
                    data, 0, data.length);            
            int degrees = getCameraDisplayOrientation(root_act); // 後述のメソッド
            Matrix m = new Matrix();
            
            if(isInCamMode){
            	m.postRotate(270);
            }else{
                m.postRotate(degrees);            	
            }

            Bitmap rotatedBitmap = Bitmap.createBitmap(origBitmap, 0, 0, origBitmap.getWidth(), origBitmap.getHeight(), m, false);
            
			// ---
			// もし、画像が大きかったら縮小して読み込む
			// 今回はimageSizeMaxの大きさに合わせる
			Bitmap shrinked_bitmap;
			int imageSizeMax = 240;
			float imageScaleWidth = rotatedBitmap.getWidth() / imageSizeMax;
			float imageScaleHeight = rotatedBitmap.getHeight() / imageSizeMax;

			// もしも、縮小できるサイズならば、縮小して読み込む
			if (imageScaleWidth > 2 && imageScaleHeight > 2) {
				BitmapFactory.Options imageOptions2 = new BitmapFactory.Options();

				// 縦横、小さい方に縮小するスケールを合わせる
				int imageScale = (int) Math
						.floor((imageScaleWidth > imageScaleHeight ? imageScaleHeight : imageScaleWidth));

				// inSampleSizeには2のべき上が入るべきなので、imageScaleに最も近く、かつそれ以下の2のべき上の数を探す
				for (int i = 2; i <= imageScale; i *= 2) {
					imageOptions2.inSampleSize = i;
				}

				ByteArrayOutputStream baos = new ByteArrayOutputStream();
				rotatedBitmap.compress(CompressFormat.JPEG, 100, baos);
				byte[] rot_bytes = baos.toByteArray();
				ByteArrayInputStream bis = new ByteArrayInputStream(rot_bytes);
				shrinked_bitmap = BitmapFactory.decodeStream(bis, null, imageOptions2);
				try {
					bis.close();
				} catch (IOException e) {
					e.printStackTrace();
				}				
				System.out.println("Sample Size: 1/" + imageOptions2.inSampleSize);				
			} else {
				shrinked_bitmap = rotatedBitmap;
			}
			// ---
                                                
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            shrinked_bitmap.compress(CompressFormat.JPEG, 100, baos);
            byte[] shrinked_data = baos.toByteArray();
            
            //String saveDir = Environment.getExternalStorageDirectory().getPath() + "/GanmenScouter";
			File saveDir = GanmenScouter.curentInstance.getFilesDir();

			/*
            // SD カードフォルダを取得
            File file = new File(saveDir);

            // フォルダ作成
            if (!file.exists()) {
                if (!file.mkdir()) {
                    System.out.println("error: mkdir failed");
                }
            }
			 */

            // 画像保存パス
            Calendar cal = Calendar.getInstance();
            SimpleDateFormat sf = new SimpleDateFormat("yyyyMMdd_HHmmss");
            //String imgPath = saveDir + "/" + sf.format(cal.getTime()) + ".jpg";
			String imgPath = saveDir.getPath() + "/" + sf.format(cal.getTime()) + ".jpg";

            // ファイル保存
            FileOutputStream fos;
            try {
                fos = new FileOutputStream(imgPath, true);
                fos.write(shrinked_data);
                fos.close();

                // アンドロイドのデータベースへ登録
                // (登録しないとギャラリーなどにすぐに反映されないため)
                registAndroidDB(imgPath);

            } catch (Exception e) {
                System.out.println(e);
            }

            fos = null;

            double tmp = 0;
            if(!isMaleMode){
            	tmp = facePP.measure_similarity(facePP.get_face_id("japanese_bijin.png", getAssets()), facePP.get_face_id(shrinked_data));
            }else{
            	tmp = facePP.measure_similarity(facePP.get_face_id("otoko_ikemen.png", getAssets()), facePP.get_face_id(shrinked_data));
            }
            
            double result_val = 50 + tmp;
            result_val = Math.floor(result_val);
//            tv_top.setText(String.valueOf(result_val) + "点");
            
            mIsTake = false;
 
            // takePicture するとプレビューが停止するので、再度プレビュースタート
            if(isInCamMode){
            	inCam.startPreview();
            }else{
            	mCam.startPreview();             	
            }
            
            try {
            	called_intent = true;
                Intent intent = new Intent();
                intent.setAction(Intent.ACTION_SEND);
                intent.setType("text/plain");
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                intent.putExtra(Intent.EXTRA_TEXT, "顔面偏差値 " + String.valueOf(result_val) + "点でした! http://bit.ly/1NbvhcO  #顔面スカウター");
                startActivityForResult(Intent.createChooser(intent, String.valueOf(result_val) + "点を共有"), 101);
            } catch (Exception e) {
            }

            if (mInterstitialAd.isLoaded()) {
                mInterstitialAd.show();
            }
        }
    };
    
    /**
     * アンドロイドのデータベースへ画像のパスを登録
     * @param path 登録するパス
     */
    private void registAndroidDB(String path) {
        // アンドロイドのデータベースへ登録
        // (登録しないとギャラリーなどにすぐに反映されないため)
        ContentValues values = new ContentValues();
        ContentResolver contentResolver = GanmenScouter.this.getContentResolver();
        values.put(Images.Media.MIME_TYPE, "image/jpeg");
        values.put("_data", path);
        contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values);
    }
    
	@Override
	protected void onResume() {
		super.onResume();
	    
		if(isInCamMode){
			inCam.startPreview();
		}else{
			mCam.startPreview();		
		}
	}

    @Override
    protected void onPause() {
        super.onPause();    	
    }
    
    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    private static int getCameraDisplayOrientation(Activity activity) {
        int rotation = activity.getWindowManager().getDefaultDisplay().getRotation();
        int degrees = 0;
        switch (rotation) {
        case Surface.ROTATION_0:
            degrees = 0;
            break;
        case Surface.ROTATION_90:
            degrees = 90;
            break;
        case Surface.ROTATION_180:
            degrees = 180;
            break;
        case Surface.ROTATION_270:
            degrees = 270;
            break;
        }
        return (90 + 360 - degrees) % 360;
    }
}

File: ganmenScouter_android/src/main/java/com/ganmen/GanmenScouter.kt
package com.ganmen

import android.Manifest
import android.app.Activity
import android.app.AlertDialog
import android.content.ContentValues
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Bitmap.CompressFormat
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.Matrix
import android.hardware.Camera
import android.hardware.Camera.*
import android.os.Bundle
import android.provider.MediaStore.Images
import android.view.MotionEvent
import android.view.Surface
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.ganmen.GanmenScouter
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.function.Supplier

/*
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.InterstitialAd;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.initialization.InitializationStatus;
import com.google.android.gms.ads.initialization.OnInitializationCompleteListener;
*/
class GanmenScouter : Activity() {
    private val WRAP_CONTENT = ViewGroup.LayoutParams.WRAP_CONTENT
    private var mchange_btn: Button? = null
    private var incam_btn: Button? = null
    private var tv_top: TextView? = null

    // カメラインスタンス
    private var mCam: Camera? = null
    private var inCam: Camera? = null

    // カメラプレビュークラス
    private var mCamPreview: CameraPreview? = null

    // 画面タッチの2度押し禁止用フラグ
    private var mIsTake = false
    private var isMaleMode = false
    private var isInCamMode = false
    var rootView_: View? = null
    var root_act: Activity = this
    var called_intent = false

    //InterstitialAd mInterstitialAd;
    var preview: RelativeLayout? = null
    var facePP = FacePPService()
    private var isInitializedUI = false
    private fun hasInCam(): Boolean {
        val numberOfCameras = getNumberOfCameras()
        return if (numberOfCameras >= 2) {
            true
        } else {
            false
        }
    }

    public override fun onStart() {
        super.onStart()
        curentInstance = this
    }

    private val permissions: Unit
        private get() {
            if (ContextCompat.checkSelfPermission(this,
                            Manifest.permission.CAMERA)
                    != PackageManager.PERMISSION_GRANTED) {
                if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                                Manifest.permission.CAMERA)) {
                    AlertDialog.Builder(this).setMessage("端末の「設定」でカメラの権限を許可してください。")
                            .setPositiveButton("OK") { dialog, id -> }
                            .setNegativeButton("Cancel") { dialog, id -> permissions }.create().show()
                } else {
                    ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA), REQUEST_CODE)
                }
            } else {
            }
            if (ContextCompat.checkSelfPermission(this,
                            Manifest.permission.WRITE_EXTERNAL_STORAGE)
                    != PackageManager.PERMISSION_GRANTED) {
                if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                                Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
                    AlertDialog.Builder(this).setMessage("端末の「設定」でストレージへの書き込みの権限を許可してください。")
                            .setPositiveButton("OK") { dialog, id -> }
                            .setNegativeButton("Cancel") { dialog, id -> permissions }.create().show()
                } else {
                    ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), REQUEST_CODE)
                }
            } else {
            }
        }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)
        permissions

/*
		MobileAds.initialize(this, new OnInitializationCompleteListener() {
			@Override
			public void onInitializationComplete(InitializationStatus initializationStatus) {
				//requestNewInterstitial();
			}
		});

		mInterstitialAd = new InterstitialAd(this);
		mInterstitialAd.setAdUnitId("ca-app-pub-3869533485696941/9899777318");
                requestNewInterstitial();

		mInterstitialAd.setAdListener(new AdListener() {
			@Override
			public void onAdClosed() {
				requestNewInterstitial();
				setup_cam_and_preview(false);
			}
		});
 */
        setup_cam_and_preview()
    }

    /*
    private void requestNewInterstitial() {
        mInterstitialAd.loadAd(new AdRequest.Builder().build());
    }
 */
    private fun setup_cam_and_preview() {
        if (true.also { isInitializedUI = it }) {
            val preview = findViewById<View>(R.id.cameraPreview) as RelativeLayout
            preview.removeView(mCamPreview)
        }
        isInitializedUI = true
        while (ContextCompat.checkSelfPermission(this,
                        Manifest.permission.CAMERA)
                != PackageManager.PERMISSION_GRANTED) {
            try {
                Thread.sleep(500)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
        // カメラインスタンスの取得
        try {
            if (mCam != null) {
                mCam!!.stopPreview()
                mCam!!.release()
                mCam = null
            }
            if (inCam != null) {
                inCam!!.stopPreview()
                inCam!!.release()
                inCam = null
            }
            if (isInCamMode && inCam == null) {
                inCam = open(1)
            } else if (mCam == null) {
                mCam = open(0)
            }
        } catch (e: Exception) {
            // エラー
            finish()
        }
        val currentCam = if (isInCamMode) inCam else mCam
        val params = currentCam!!.parameters
        val min_width = -1
        //Picture解像度取得
        val supportedPictureSizes = params.supportedPictureSizes
        if (supportedPictureSizes != null && supportedPictureSizes.size > 1) {
            for (i in supportedPictureSizes.indices) {
                val size = supportedPictureSizes[i]
                //解像度表示
                println("picture width:" + size.width + " height:" + size.height)
                if (min_width == -1) {
                    width = size.width
                    height = size.height
                } else if (size.width < width) {
                    width = size.width
                    height = size.height
                }
            }
        }
        params.setPictureSize(width, height)
        params.setPreviewSize(width, height)
        val preview = findViewById<View>(R.id.cameraPreview) as RelativeLayout
        mCamPreview = CameraPreview(this, currentCam)
        setCameraDisplayOrientation(this, if (isInCamMode) 0 else 1, currentCam)
        preview.addView(mCamPreview)

        // mCamPreview に タッチイベントを設定
        mCamPreview!!.setOnTouchListener { v, event ->
            if (event.action == MotionEvent.ACTION_DOWN) {
                if (!mIsTake) {
                    // 撮影中の2度押し禁止用フラグ
                    mIsTake = true
                    currentCam?.autoFocus(autoFocusListener_)
                    // 画像取得
                    // mCam.takePicture(null, null, mPicJpgListener);
                }
            }
            true
        }
        tv_top = TextView(this)
        tv_top!!.setTextColor(Color.RED)
        tv_top!!.text = "　　　　　　　　　　　　　　　画面タッチで測定!"
        preview.addView(tv_top, FrameLayout.LayoutParams.WRAP_CONTENT)
        val layout_buttons = LinearLayout(this)
        layout_buttons.orientation = LinearLayout.VERTICAL
        if (hasInCam()) {
            incam_btn = Button(this)
            if (isInCamMode) {
                incam_btn!!.text = "メインカメラモードへ切替"
            } else {
                incam_btn!!.text = "インカムモードへ切替"
            }
            incam_btn!!.setOnClickListener {
                if (isInCamMode) {
                    isInCamMode = false
                    setup_cam_and_preview()
                } else {
                    isInCamMode = true
                    setup_cam_and_preview()
                }
            }
        }
        layout_buttons.addView(incam_btn)
        mchange_btn = Button(this)
        mchange_btn!!.text = "男性撮影モードへ切替"
        mchange_btn!!.setOnClickListener {
            if (isMaleMode == false) {
                isMaleMode = true
                mchange_btn!!.text = "女性撮影モードへ切替"
            } else {
                isMaleMode = false
                mchange_btn!!.text = "男性撮影モードへ切替"
            }
        }
        layout_buttons.addView(mchange_btn)
        preview.addView(layout_buttons)
    }

    // AF完了時のコールバック
    private val autoFocusListener_ = AutoFocusCallback { success, camera ->
        camera.autoFocus(null)
        camera.takePicture(null, null, mPicJpgListener)
    }

    /**
     * JPEG データ生成完了時のコールバック
     */
    private val mPicJpgListener = PictureCallback { data, camera ->
        if (data == null) {
            return@PictureCallback
        }
        val origBitmap = BitmapFactory.decodeByteArray(
                data, 0, data.size)
        val degrees = getCameraDisplayOrientation(root_act) // 後述のメソッド
        val m = Matrix()
        if (isInCamMode) {
            m.postRotate(270f)
        } else {
            m.postRotate(degrees.toFloat())
        }
        val rotatedBitmap = Bitmap.createBitmap(origBitmap, 0, 0, origBitmap.width, origBitmap.height, m, false)

        // ---
        // もし、画像が大きかったら縮小して読み込む
        // 今回はimageSizeMaxの大きさに合わせる
        val shrinked_bitmap: Bitmap?
        val imageSizeMax = 240
        val imageScaleWidth = rotatedBitmap.width / imageSizeMax.toFloat()
        val imageScaleHeight = rotatedBitmap.height / imageSizeMax.toFloat()

        // もしも、縮小できるサイズならば、縮小して読み込む
        if (imageScaleWidth > 2 && imageScaleHeight > 2) {
            val imageOptions2 = BitmapFactory.Options()

            // 縦横、小さい方に縮小するスケールを合わせる
            val imageScale = Math
                    .floor((if (imageScaleWidth > imageScaleHeight) imageScaleHeight else imageScaleWidth).toDouble()).toInt()

            // inSampleSizeには2のべき上が入るべきなので、imageScaleに最も近く、かつそれ以下の2のべき上の数を探す
            var i = 2
            while (i <= imageScale) {
                imageOptions2.inSampleSize = i
                i *= 2
            }
            val baos = ByteArrayOutputStream()
            rotatedBitmap.compress(CompressFormat.JPEG, 100, baos)
            val rot_bytes = baos.toByteArray()
            val bis = ByteArrayInputStream(rot_bytes)
            shrinked_bitmap = BitmapFactory.decodeStream(bis, null, imageOptions2)
            try {
                bis.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            println("Sample Size: 1/" + imageOptions2.inSampleSize)
        } else {
            shrinked_bitmap = rotatedBitmap
        }
        // ---
        val baos = ByteArrayOutputStream()
        shrinked_bitmap!!.compress(CompressFormat.JPEG, 100, baos)
        val shrinked_data = baos.toByteArray()
        val saveDir = curentInstance!!.filesDir

        // 画像保存パス
        val cal = Calendar.getInstance()
        val sf = SimpleDateFormat("yyyyMMdd_HHmmss")
        val imgPath = saveDir.path + "/" + sf.format(cal.time) + ".jpg"

        // ファイル保存
        var fos: FileOutputStream?
        try {
            fos = FileOutputStream(imgPath, true)
            fos.write(shrinked_data)
            fos.close()

            // アンドロイドのデータベースへ登録
            // (登録しないとギャラリーなどにすぐに反映されないため)
            registAndroidDB(imgPath)
        } catch (e: Exception) {
            println(e)
        }
        fos = null
        if (!isMaleMode) {
            val sup = Supplier {
                val tmp = facePP.measure_similarity(facePP.get_face_id("japanese_bijin.png", assets), facePP.get_face_id(shrinked_data))
                doubleTaskResult = tmp
                tmp
            }
            val future = CompletableFuture.supplyAsync(sup)
            try {
                Thread.sleep(7000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        } else {
            val sup = Supplier {
                val tmp = facePP.measure_similarity(facePP.get_face_id("otoko_ikemen.png", assets), facePP.get_face_id(shrinked_data))
                doubleTaskResult = tmp
                tmp
            }
            val future = CompletableFuture.supplyAsync(sup)
            try {
                Thread.sleep(7000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
        var result_val = 50.0 + doubleTaskResult
        result_val = Math.floor(result_val)
        mIsTake = false

        // takePicture するとプレビューが停止するので、再度プレビュースタート
        camera.startPreview()
        try {
            called_intent = true
            val intent = Intent()
            intent.action = Intent.ACTION_SEND
            intent.type = "text/plain"
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            intent.putExtra(Intent.EXTRA_TEXT, "顔面偏差値 " + result_val.toString() + "点でした! http://bit.ly/1NbvhcO  #顔面スカウター")
            startActivityForResult(Intent.createChooser(intent, result_val.toString() + "点を共有"), 101)
        } catch (e: Exception) {
        }
        /*
        if (mInterstitialAd.isLoaded()) {
            mInterstitialAd.show();
        }
*/
    }

    /**
     * アンドロイドのデータベースへ画像のパスを登録
     * @param path 登録するパス
     */
    private fun registAndroidDB(path: String) {
        // アンドロイドのデータベースへ登録
        // (登録しないとギャラリーなどにすぐに反映されないため)
        val values = ContentValues()
        val contentResolver = this@GanmenScouter.contentResolver
        values.put(Images.Media.MIME_TYPE, "image/jpeg")
        values.put("_data", path)
        contentResolver.insert(Images.Media.EXTERNAL_CONTENT_URI, values)
    }

    override fun onResume() {
        super.onResume()
        if (isInCamMode) {
            inCam!!.startPreview()
        } else {
            mCam!!.startPreview()
        }
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    companion object {
        const val REQUEST_CODE = 1

        // プレビューサイズ
        private var width = -1
        private var height = -1
        var doubleTaskResult = 0.0
        var stringTaskResult: String? = null
        var curentInstance: GanmenScouter? = null
        fun setCameraDisplayOrientation(activity: Activity,
                                        cameraId: Int, camera: Camera?) {
            val info = CameraInfo()
            getCameraInfo(cameraId, info)
            val rotation = activity.windowManager.defaultDisplay
                    .rotation
            var degrees = 0
            when (rotation) {
                Surface.ROTATION_0 -> degrees = 0
                Surface.ROTATION_90 -> degrees = 90
                Surface.ROTATION_180 -> degrees = 180
                Surface.ROTATION_270 -> degrees = 270
            }
            var result: Int
            if (info.facing == CameraInfo.CAMERA_FACING_FRONT) {
                result = (info.orientation + degrees) % 360
                result = (360 - result) % 360 // compensate the mirror
            } else {  // back-facing
                result = (info.orientation - degrees + 360) % 360
            }
            camera!!.setDisplayOrientation(result)
        }

        private fun getCameraDisplayOrientation(activity: Activity): Int {
            val rotation = activity.windowManager.defaultDisplay.rotation
            var degrees = 0
            when (rotation) {
                Surface.ROTATION_0 -> degrees = 0
                Surface.ROTATION_90 -> degrees = 90
                Surface.ROTATION_180 -> degrees = 180
                Surface.ROTATION_270 -> degrees = 270
            }
            return (90 + 360 - degrees) % 360
        }
    }
}